db/go_level_db.go

```diff
@@ -169,102 +162,104 @@ func (mBatch *goLevelDBBatch) Write() {
 
 //----------------------------------------
 // Iterator
+// NOTE This is almost identical to db/c_level_db.Iterator
+// Before creating a third version, refactor.
+
+type goLevelDBIterator struct {
+	source    iterator.Iterator
+	start     []byte
+	end       []byte
+	isReverse bool
+	isInvalid bool
+}
+
+var _ Iterator = (*goLevelDBIterator)(nil)
 
-// https://godoc.org/github.com/syndtr/goleveldb/leveldb#DB.NewIterator
-// A nil Range.Start is treated as a key before all keys in the DB.
-// And a nil Range.Limit is treated as a key after all keys in the DB.
-func goLevelDBIterRange(start, end []byte) *util.Range {
-	// XXX: what if start == nil ?
-	if len(start) == 0 {
-		start = nil
+func newGoLevelDBIterator(source iterator.Iterator, start, end []byte, isReverse bool) *goLevelDBIterator {
+	if isReverse {
+		panic("not implemented yet") // XXX
 	}
-	return &util.Range{
-		Start: start,
-		Limit: end,
+	source.Seek(start)
+	return &goLevelDBIterator{
+		source:    source,
+		start:     start,
+		end:       end,
+		isReverse: false,
+		isInvalid: false,
 	}
 }
 
 func (db *GoLevelDB) Iterator(start, end []byte) Iterator {
-	itrRange := goLevelDBIterRange(start, end)
-	itr := db.db.NewIterator(itrRange, nil)
-	itr.Seek(start) // if we don't call this the itr is never valid (?!)
-	return &goLevelDBIterator{
-		source: itr,
-		start:  start,
-		end:    end,
-	}
+	itr := db.db.NewIterator(nil, nil)
+	return newGoLevelDBIterator(itr, start, end, false)
 }
 
 func (db *GoLevelDB) ReverseIterator(start, end []byte) Iterator {
-	// XXX
-	return nil
+	panic("not implemented yet") // XXX
 }
 
-var _ Iterator = (*goLevelDBIterator)(nil)
-
-type goLevelDBIterator struct {
-	source     iterator.Iterator
-	invalid    bool
-	start, end []byte
+func (itr *goLevelDBIterator) Domain() ([]byte, []byte) {
+	return itr.start, itr.end
 }
 
-func (it *goLevelDBIterator) Domain() ([]byte, []byte) {
-	return it.start, it.end
-}
+func (itr *goLevelDBIterator) Valid() bool {
 
-// Key returns a copy of the current key.
-func (it *goLevelDBIterator) Key() []byte {
-	if !it.Valid() {
-		panic("goLevelDBIterator Key() called when invalid")
+	// Once invalid, forever invalid.
+	if itr.isInvalid {
+		return false
 	}
-	key := it.source.Key()
-	k := make([]byte, len(key))
-	copy(k, key)
 
-	return k
-}
+	// Panic on DB error.  No way to recover.
+	itr.assertNoError()
 
-// Value returns a copy of the current value.
-func (it *goLevelDBIterator) Value() []byte {
-	if !it.Valid() {
-		panic("goLevelDBIterator Value() called when invalid")
+	// If source is invalid, invalid.
+	if !itr.source.Valid() {
+		itr.isInvalid = true
+		return false
 	}
-	val := it.source.Value()
-	v := make([]byte, len(val))
-	copy(v, val)
-
-	return v
-}
 
-func (it *goLevelDBIterator) Valid() bool {
-	it.assertNoError()
-	if it.invalid {
+	// If key is end or past it, invalid.
```

Comment:
technically we don't need to do this here since go-leveldb lets you specify the range in the iterator.

but for consistency with c-leveldb, ok